osu.Game/Rulesets/Mods/ModAdaptiveSpeed.cs

```diff
@@ -0,0 +1,183 @@
+
+        /// <summary>
+        /// Adjust track rate using the average speed of the last x hits
+        /// </summary>
+        private const int average_count = 6;
+
+        public override string Name => "Adaptive Speed";
+
+        public override string Acronym => "AS";
+
+        public override string Description => "Let track speed adapt to you.";
+
+        public override ModType Type => ModType.Fun;
+
+        public override double ScoreMultiplier => 1;
+
+        public override Type[] IncompatibleMods => new[] { typeof(ModRateAdjust), typeof(ModTimeRamp) };
+
+        [SettingSource("Initial rate", "The starting speed of the track")]
+        public BindableNumber<double> InitialRate { get; } = new BindableDouble
+        {
+            MinValue = 0.5,
+            MaxValue = 2,
+            Default = 1,
+            Value = 1,
+            Precision = 0.01
+        };
+
+        [SettingSource("Adjust pitch", "Should pitch be adjusted with speed")]
+        public BindableBool AdjustPitch { get; } = new BindableBool
+        {
+            Default = true,
+            Value = true
+        };
+
+        public BindableNumber<double> SpeedChange { get; } = new BindableDouble
+        {
+            Default = 1,
+            Value = 1,
+            Precision = 0.01
+        };
+
+        private ITrack track;
+        private HUDOverlay overlay;
+
+        private readonly List<double> recentRates = Enumerable.Range(0, average_count).Select(_ => 1d).ToList();
+
+        // rate for a hit is calculated using the end time of another hit object earlier in time
+        // caching them here for easy access
+        private readonly Dictionary<HitObject, double> previousEndTimes = new Dictionary<HitObject, double>();
+
+        // record the value removed from recentRates when an object is hit
+        // for rewind support
+        private readonly Dictionary<HitObject, double> dequeuedRates = new Dictionary<HitObject, double>();
+
+        public ModAdaptiveSpeed()
+        {
+            InitialRate.BindValueChanged(val => SpeedChange.Value = val.NewValue);
+            AdjustPitch.BindValueChanged(applyPitchAdjustment);
+        }
+
+        public void ApplyToHUD(HUDOverlay overlay)
+        {
+            // this is only used to transform the SpeedChange bindable
+            this.overlay = overlay;
+        }
+
+        public void ApplyToTrack(ITrack track)
+        {
+            this.track = track;
+
+            InitialRate.TriggerChange();
+            AdjustPitch.TriggerChange();
+            recentRates.Clear();
+            recentRates.AddRange(Enumerable.Range(0, average_count).Select(_ => InitialRate.Value));
+        }
+
+        public void ApplyToSample(DrawableSample sample)
+        {
+            sample.AddAdjustment(AdjustableProperty.Frequency, SpeedChange);
+        }
+
+        public double ApplyToRate(double time, double rate = 1) => rate * InitialRate.Value;
```

Comment:
From a difficulty calculation standpoint, this should be `rate * slowest_rate` I'm pretty sure to match worst case scenarios. From testing the mod out, it's pretty feasible to slow down the map to very low rates whilst hitting all 300s (if you hit late on every note)

Reply:
Difficulty calculation uses `ApplyToTrack` instead of `ApplyToRate` for rate determination, and it's not a good idea to use `slowest_rate` in `ApplyToTrack` because song select plays at that rate too.

Reply:
Ah, that's down to me misremembering - since I used `ApplyToRate` in `ClockWithMods` in https://github.com/ppy/osu/pull/14831. Nonetheless, I'm pretty confident in my assessment on difficulty with the current proposed score structure (unless score data somehow starts including state of adaptive mods like this one) - maybe it's too early to action upon it in this PR though.

Reply:
I'd say the diffcalc implications don't need to be addressed at this stage yet. Other variable rate mods are already handled on a best-effort basis. Would prefer to keep this as is for now.